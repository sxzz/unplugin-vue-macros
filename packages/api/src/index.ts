export { MagicString } from '@vue-macros/common'

// WIP
// export * from './vue'
export * from './ts'
export * from './utils'
