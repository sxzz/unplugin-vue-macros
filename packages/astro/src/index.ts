import { type AstroIntegration } from 'astro'
import { type Options } from 'unplugin-vue-macros'
import VitePlugin from 'unplugin-vue-macros/vite'

const name = 'unplugin-vue-macros'

export default function VueMacrosAstroIntegration(
  userOptions: Options = {}
): AstroIntegration {
  return {
    name,
    hooks: {
      'astro:config:setup': ({ updateConfig }) => {
        updateConfig({
          vitePlugin: [VitePlugin(userOptions)],
        })
      },
    },
  }
}
