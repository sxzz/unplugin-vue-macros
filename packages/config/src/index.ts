export * from './config-worker'
export * from './config'
export * from './define'
export * from './options'
