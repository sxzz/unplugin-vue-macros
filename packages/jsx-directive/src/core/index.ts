import {
  MagicString,
  REGEX_SETUP_SFC,
  babelParse,
  generateTransform,
  getLang,
  parseSFC,
  walkAST,
} from '@vue-macros/common'
import { transformVIf } from './v-if'
import { transformVFor } from './v-for'
import { transformVMemo } from './v-memo'
import { transformVHtml } from './v-html'
import { type VSlotMap, transformVSlot } from './v-slot'
import { transformVOn, transformVOnWithModifiers } from './v-on'
import type { JSXAttribute, JSXElement, Node, Program } from '@babel/types'

export type JsxDirective = {
  node: JSXElement
  attribute: JSXAttribute
  parent?: Node | null
  vForAttribute?: JSXAttribute
  vMemoAttribute?: JSXAttribute
}

export function transformJsxDirective(
  code: string,
  id: string,
  version: number,
) {
  const lang = getLang(id)
  let asts: {
    ast: Program
    offset: number
  }[] = []
  if (lang === 'vue' || REGEX_SETUP_SFC.test(id)) {
    const { scriptSetup, getSetupAst, script, getScriptAst } = parseSFC(
      code,
      id,
    )
    if (script) {
      asts.push({ ast: getScriptAst()!, offset: script.loc.start.offset })
    }
    if (scriptSetup) {
      asts.push({ ast: getSetupAst()!, offset: scriptSetup.loc.start.offset })
    }
  } else if (['jsx', 'tsx'].includes(lang)) {
    asts = [{ ast: babelParse(code, lang), offset: 0 }]
  } else {
    return
  }

  const s = new MagicString(code)
  for (const { ast, offset } of asts) {
    const vIfMap = new Map<Node | null | undefined, JsxDirective[]>()
    const vForNodes: JsxDirective[] = []
    const vMemoNodes: (JsxDirective & {
      vForAttribute?: JSXAttribute
    })[] = []
    const vHtmlNodes: JsxDirective[] = []
    const vSlotMap: VSlotMap = new Map()
    const vOnNodes: JsxDirective[] = []
    const vOnWithModifiers: JsxDirective[] = []
    walkAST<Node>(ast, {
      enter(node, parent) {
        if (node.type !== 'JSXElement') return
        const tagName = s.sliceNode(node.openingElement.name, {
          offset,
        })

        let vIfAttribute
        let vForAttribute
        let vMemoAttribute
        let vSlotAttribute
        for (const attribute of node.openingElement.attributes) {
          if (attribute.type !== 'JSXAttribute') continue

          if (
            ['v-if', 'v-else-if', 'v-else'].includes(`${attribute.name.name}`)
          ) {
            vIfAttribute = attribute
          }

          if (attribute.name.name === 'v-for') {
            vForAttribute = attribute
          }

          if (['v-memo', 'v-once'].includes(`${attribute.name.name}`)) {
            vMemoAttribute = attribute
          }

          if (attribute.name.name === 'v-html') {
            vHtmlNodes.push({
              node,
              attribute,
            })
          }

          if (
            (attribute.name.type === 'JSXNamespacedName'
              ? attribute.name.namespace
              : attribute.name
            ).name === 'v-slot'
          ) {
            vSlotAttribute = attribute
          }

          if (attribute.name.name === 'v-on') {
            vOnNodes.push({
              node,
              attribute,
            })
          }

          if (/^on[A-Z]\S*_\S+/.test(`${attribute.name.name}`)) {
            vOnWithModifiers.push({
              node,
              attribute,
            })
          }
        }

        if (vIfAttribute && !(vSlotAttribute && tagName === 'template')) {
          vIfMap.get(parent) || vIfMap.set(parent, [])
          vIfMap.get(parent)!.push({
            node,
            attribute: vIfAttribute,
            parent,
          })
        }

        if (vForAttribute) {
          vForNodes.push({
            node,
            attribute: vForAttribute,
            parent: vIfAttribute ? undefined : parent,
            vMemoAttribute,
          })
        }

        if (vMemoAttribute) {
          vMemoNodes.push({
            node,
            attribute: vMemoAttribute,
            parent: vForAttribute || vIfAttribute ? undefined : parent,
            vForAttribute,
          })
        }

        if (vSlotAttribute) {
          const slotNode = tagName === 'template' ? parent : node
          if (slotNode?.type !== 'JSXElement') return

          const attributeMap =
            vSlotMap.get(slotNode)?.attributeMap ||
            vSlotMap
              .set(slotNode, {
                vSlotAttribute:
                  tagName !== 'template' ? vSlotAttribute : undefined,
                attributeMap: new Map(),
              })
              .get(slotNode)!.attributeMap
          const children =
            attributeMap.get(vSlotAttribute)?.children ||
            attributeMap
              .set(vSlotAttribute, {
                children: [],
                vIfAttribute:
                  tagName === 'template' && vIfAttribute
                    ? vIfAttribute
                    : undefined,
              })
              .get(vSlotAttribute)!.children

          if (slotNode === parent) {
            children.push(node)

            if (attributeMap.get(null)) return
            for (const child of parent.children) {
              if (
                (child.type === 'JSXElement' &&
                  s.sliceNode(child.openingElement.name, { offset }) ===
                    'template') ||
                (child.type === 'JSXText' &&
                  !s.sliceNode(child, { offset }).trim())
              )
                continue
              const defaultNodes =
                attributeMap.get(null)?.children ||
                attributeMap.set(null, { children: [] }).get(null)!.children
              defaultNodes.push(child)
            }
          } else {
            children.push(...node.children)
          }
        }
      },
    })

    vIfMap.forEach((nodes) => transformVIf(nodes, s, offset, version))
    transformVFor(vForNodes, s, offset, version)
    version >= 3.2 && transformVMemo(vMemoNodes, s, offset)
    transformVHtml(vHtmlNodes, s, offset, version)
    transformVOn(vOnNodes, s, offset, version)
    transformVOnWithModifiers(vOnWithModifiers, s, offset, version)
    transformVSlot(vSlotMap, s, offset, version)
  }

  return generateTransform(s, id)
}
