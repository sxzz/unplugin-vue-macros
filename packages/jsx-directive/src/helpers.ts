export { renderList, unref, withKeys, withMemo, withModifiers } from 'vue'

export { createPropsDefaultProxy } from './core'
