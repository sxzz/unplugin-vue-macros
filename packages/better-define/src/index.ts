import process from 'node:process'
import { RollupResolve, setResolveTSFileIdImpl } from '@vue-macros/api'
import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { transformBetterDefine } from './core'
import type { PluginContext } from 'rollup'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    const { resolve, handleHotUpdate } = RollupResolve()

    return {
      name,
      enforce: 'pre',

      buildStart() {
        if (framework === 'rollup' || framework === 'vite') {
          setResolveTSFileIdImpl(resolve(this as PluginContext))
        }
      },

      transformInclude: filter,
      async transform(code, id) {
        try {
          return await transformBetterDefine(code, id, options.isProduction)
        } catch (error: unknown) {
          this.warn(`${name} ${error}`)
          console.warn(error)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction ??= config.isProduction
        },

        handleHotUpdate,
      },
    }
  },
)
export default plugin
