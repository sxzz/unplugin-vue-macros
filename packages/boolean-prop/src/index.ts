import { getVuePluginApi, type VuePluginApi } from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { transformBooleanProp, type Options } from './core/index'
import type { Plugin } from 'vite'

// legacy export
export * from './api'

const name = generatePluginName()

function rollup(options: Options = {}): Plugin {
  let api: VuePluginApi | null | undefined

  return {
    name,
    configResolved(config) {
      try {
        api = getVuePluginApi(config.plugins)
      } catch {}
    },
    buildStart(rollupOpts) {
      if (api === undefined)
        try {
          api = getVuePluginApi(rollupOpts.plugins)
        } catch (error: any) {
          this.warn(error)
          return
        }

      if (!api) return

      api.options.template ||= {}
      api.options.template.compilerOptions ||= {}
      api.options.template.compilerOptions.nodeTransforms ||= []

      api.options.template.compilerOptions.nodeTransforms.push(
        transformBooleanProp(options),
      )
    },
  }
}

const plugin: {
  rollup: typeof rollup
  vite: typeof rollup
} = {
  rollup,
  vite: rollup,
}
export default plugin
