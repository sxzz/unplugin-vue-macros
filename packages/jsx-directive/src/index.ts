import { type UnpluginContextMeta, createUnplugin } from 'unplugin'
import {
  type BaseOptions,
  FilterFileType,
  type MarkRequired,
  REGEX_NODE_MODULES,
  createFilter,
  detectVueVersion,
  getFilterPattern,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' assert { type: 'macro' }
import { transformJsxDirective } from './core'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<Options, 'version'>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SRC_FILE],
    framework,
  )
  return {
    include,
    exclude: [REGEX_NODE_MODULES],
    ...options,
    version,
  }
}

const name = generatePluginName()

export default createUnplugin<Options | undefined, false>(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      transformInclude: filter,
      transform(code, id) {
        return transformJsxDirective(code, id, options.version)
      },
    }
  },
)
