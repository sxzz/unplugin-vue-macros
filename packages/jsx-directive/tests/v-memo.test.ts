import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformJsxDirective } from '../src/api'

describe('jsx-vue-directive', () => {
  describe('v-memo', async () => {
    await testFixtures(
      import.meta.glob<string>('./fixtures/v-memo/*.{vue,jsx,tsx}', {
        eager: true,
        query: '?raw',
        import: 'default',
      }),
      (_, id, code) => transformJsxDirective(code, id, 3.2)?.code,
    )
  })
})
