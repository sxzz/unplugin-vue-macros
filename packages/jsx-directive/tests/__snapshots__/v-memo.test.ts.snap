// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsx-vue-directive > v-memo > ./fixtures/v-memo/v-memo.setup.tsx 1`] = `
"
import { renderList as __MACROS_renderList } from "@vue-macros/jsx-directive/helpers";
import { withMemo as __MACROS_withMemo } from "@vue-macros/jsx-directive/helpers";const __MACROS_cache = [];__MACROS_cache[0] = [];// @ts-nocheck
const list = [1, 2, 3]
let selected = $ref(0)

export default (
  <>
    {__MACROS_renderList(list, (i, index) => __MACROS_withMemo([selected === i], () => <div key={index}>
      <div>
        {i}: {selected}
      </div>
    </div>, __MACROS_cache[0], 0 + index + 1))}
  </>
)
"
`;

exports[`jsx-vue-directive > v-memo > ./fixtures/v-memo/v-once.setup.tsx 1`] = `
"
import { withMemo as __MACROS_withMemo } from "@vue-macros/jsx-directive/helpers";const __MACROS_cache = [];let selected = $ref(0)

export default (
  __MACROS_withMemo([], () => <div onClick={() => selected++}>
    {selected}
  </div>, __MACROS_cache, 0)
)
"
`;
